### PR TITLE
add `any_value()` and option to enforce strict group by

### PR DESCRIPTION
**Pull request by JCOR11599**
Added support for another option in `@@sql_mode` called `ONLY_FULL_GROUP_BY`, which makes it so select statements over group by must contain columns that are part of the group by. Before, dolt was always strict about group by selects, now (to match MySQL) we are not strict by default.
Additionally, there is now support for the "aggregate" function `any_value()` which bypasses the `ONLY_FULL_GROUP_BY` if it is set.

Also, changes `GetGlobalVariable` and `GetSessionVariable` to convert `sql.SetTypes` to strings.

MySQL docs for reference:
 - `any_value()`: https://dev.mysql.com/doc/refman/8.0/en/miscellaneous-functions.html
 - `ONLY_FULL_GROUP_BY`: https://dev.mysql.com/doc/refman/5.7/en/sql-mode.html#sqlmode_only_full_group_by

fix for: https://github.com/dolthub/dolt/issues/1448
and: https://github.com/dolthub/dolt/issues/4604

### CODE REVIEW

File: sql/analyzer/validation_rules_test.go [hunk 1]
```diff
@@ -141,6 +140,8 @@ func TestValidateGroupByErr(t *testing.T) {
 		plan.NewResolvedTable(child, nil, nil),
 	)
 
+	err = sql.SystemVariables.SetGlobal("sql_mode", "STRICT_TRANS_TABLES,NO_ENGINE_SUBSTITUTION,ONLY_FULL_GROUP_BY")
```
Comment (fulghum) [inline sql/analyzer/validation_rules_test.go:143]: Should we go ahead and just make this the default for `sql_mode`? That would get us one step closer to matching MySQL's default `sql_mode`, so seems like a good change.

File: sql/analyzer/validation_rules.go [hunk 2]
```diff
@@ -211,6 +211,26 @@ func validateGroupBy(ctx *sql.Context, a *Analyzer, n sql.Node, scope *Scope, se
 	span, ctx := ctx.Span("validate_group_by")
 	defer span.End()
 
+	// only enforce strict group by when this variable is set
+	var strictGroupBy bool
+	if sysVar, sysVal, ok := sql.SystemVariables.GetGlobal("sql_mode"); ok {
```
Comment (fulghum) [inline sql/analyzer/validation_rules.go:216]: `sql_mode` can also be set for a session, so probably makes sense to have this pull from the session, instead of just the global setting.

File: enginetest/queries/script_queries.go [hunk 3]
```diff
@@ -1346,6 +1346,40 @@ var ScriptTests = []ScriptTest{
...
+			{
+				Query: "SELECT floor(cor0.col1) * ceil(cor0.col0) AS col2 FROM tab1 AS cor0 GROUP BY cor0.col0 ORDER BY cor0.col1",
+				Expected: []sql.Row{
+					{425},
+					{714},
+					{4277},
+				},
+			},
+			{
+				Query:    "SET GLOBAL sql_mode=(select concat(@@sql_mode, ',ONLY_FULL_GROUP_BY'))",
```
Comment (fulghum) [inline enginetest/queries/script_queries.go:1374]: would be nice to assert that the session var for `sql_mode` is honored, too

**Review by fulghum — COMMENTED** [review]
Very nice first pass! Mostly minor comments. I'd like to take one more look at the tests though after your next pass, just to see if I can think of any more interesting edge cases for us to hit.

File: sql/analyzer/validation_rules_test.go (hunk 1 above)
Comment (JCOR11599) [reply sql/analyzer/validation_rules_test.go:143]: Yeah this should be the default; even better since dolt already did this by default just without setting the variable.
For some reason, my MySQL server didn't have this option set, so I assumed the default was not strict.

File: sql/plan/group_by.go [hunk 4]
```diff
@@ -420,7 +440,7 @@ func newAggregationBuffer(expr sql.Expression) (sql.AggregationBuffer, error) {
 		return n.NewBuffer()
 	default:
 		// The semantics for a non-aggregation in a group by node is Last.
```
Comment (fulghum) [inline sql/plan/group_by.go:442]: Might be nice to leave a note/clue about `ONLY_FULL_GROUP_BY` to help future readers understand this. 

```suggestion
		// The semantics for a non-aggregation expression in a group by node is First.
		// When ONLY_FULL_GROUP_BY is enabled, this is an error, but it's allowed otherwise. 
```

File: sql/plan/group_by.go [hunk 5]
```diff
@@ -420,7 +440,7 @@ func newAggregationBuffer(expr sql.Expression) (sql.AggregationBuffer, error) {
...
-		return aggregation.NewLast(expr).NewBuffer()
+		return aggregation.NewFirst(expr).NewBuffer()
```
Comment (fulghum) [inline sql/plan/group_by.go:443]: Good catch!

File: sql/plan/group_by.go [hunk 6]
```diff
@@ -233,15 +233,35 @@ func (i *groupByIter) Next(ctx *sql.Context) (sql.Row, error) {
...
+	// special case for any_value
 	var err error
+	onlyAnyValue := true
 	for j, a := range i.selectedExprs {
 		i.buf[j], err = newAggregationBuffer(a)
 		if err != nil {
 			return nil, err
 		}
+		// TODO: bad, figure out better way to determine if there are any agg selections that aren't any_value
+		// maybe should be handled in analyzer?
```
Comment (fulghum) [inline sql/plan/group_by.go:245]: Good/helpful note. I think this is totally fine to special case this here like this for now.

File: sql/analyzer/validation_rules.go [hunk 7]
```diff
@@ -207,10 +207,31 @@ func validateOrderBy(ctx *sql.Context, a *Analyzer, n sql.Node, scope *Scope, se
...
 func validateGroupBy(ctx *sql.Context, a *Analyzer, n sql.Node, scope *Scope, sel RuleSelector) (sql.Node, transform.TreeIdentity, error) {
 	span, ctx := ctx.Span("validate_group_by")
 	defer span.End()
 
+	// only enforce strict group by when this variable is set
+	if isStrict, err := checkSqlMode(ctx, "ONLY_FULL_GROUP_BY"); err != nil {
+		return n, transform.SameTree, err
+	} else if !isStrict {
+		return n, transform.SameTree, nil
+	}
```
Comment (fulghum) [inline sql/analyzer/validation_rules.go:233]: oooh! This reads soooo nice now. Thanks for pulling out that helper function for `checkSqlMode`. 😎

**Review by fulghum — APPROVED** [review]
Great work James! 🙌
